src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -204,9 +205,24 @@ private void AddCompilationRoots()
                 AddCompilationRootsForRuntimeExports(module);
            }
 
+            AddCompilationRootsForBootstrap();
+
             AddCompilationRootsForRuntimeExports((EcmaModule)_typeSystemContext.SystemModule);
         }
 
+        private void AddCompilationRootsForBootstrap()
+        {
+            if (Options.IsCppCodeGen)
+                return;
```

Comment:
BTW: Everything you have here (except the string literals init) should work fine for CppCodeGen ... it is fine to clean it up in separate commit if you prefer.